OpenThread support in Zephyr ignore return values

<!--
**Notes**
Github Discussions (https://github.com/zephyrproject-rtos/zephyr/discussions)
are available to first verify that the issue is a genuine Zephyr bug and not a
consequence of Zephyr services misuse.

This issue list is only for bugs in the main Zephyr code base
(https://github.com/zephyrproject-rtos/zephyr/). If the bug is for a project
fork (such as NCS) specific feature, please open an issue in the fork project
instead.
-->

**Describe the bug**
Zephyr OpenThread usage ignores return values on OpenThread calls which requires return value to be processed.

OpenThread functions uses `otError` return type, which is defined as:
```
typedef enum OT_MUST_USE_RESULT otError
```
https://github.com/zephyrproject-rtos/openthread/blob/3ae741f95e7dfb391dec35c48742862049eb62e8/include/openthread/error.h#L56

and `OT_MUST_USE_RESULT` is used by OpenThread to define the return type should not be ignored,
> ```
> * Compiler-specific indication that a class or enum must be used when it is
> * the return value of a function.
> ```
https://github.com/zephyrproject-rtos/openthread/blob/3ae741f95e7dfb391dec35c48742862049eb62e8/include/openthread/platform/toolchain.h#L66-L81

However this is only active when compiling with clang and not available for gcc.

**To Reproduce**
Steps to reproduce the behavior:
1. Install a LLVM (clang/lld) toolchain, for example: LLVM for Arm https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm
2. Export the LLVM install dir: `export LLVM_TOOLCHAIN_PATH=<path-to-llvm>`
3. Export llvm as the toolchain to use: `export ZEPHYR_TOOLCHAIN_VARIANT=llvm`
4. Build the openthread shell sample.
   **note** depending on the LLVM toolchain installed it might be necessary to specify LLD and compiler_rt to be used (for example when using LLVM for Arm).
    ```
    $ west build -p -b nrf52840dk/nrf52840 samples/net/openthread/shell/
    ```
    alternative:
    ```
    $ west build -p -b nrf52840dk/nrf52840 samples/net/openthread/shell/ -- -DCONFIG_COMPILER_RT_RTLIB=y -DCONFIG_LLVM_USE_LLD=y
    ```
4. See the warnings (errors if `CONFIG_COMPILER_WARNINGS_AS_ERRORS=y` is used)
```
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:341:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  341 |                 otThreadSetEnabled(ot_context->instance, true);
      |                 ^~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:442:2: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  442 |         otIp6SetEnabled(ot_context->instance, true);
      |         ^~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:453:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  453 |                 otThreadSetLinkMode(ot_context->instance, ot_mode);
      |                 ^~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:454:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  454 |                 otLinkSetPollPeriod(ot_context->instance, OT_POLL_PERIOD);
      |                 ^~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:489:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  489 |                 otThreadSetNetworkName(ot_instance, OT_NETWORK_NAME);
      |                 ^~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:490:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  490 |                 otLinkSetChannel(ot_instance, OT_CHANNEL);
      |                 ^~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:491:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  491 |                 otLinkSetPanId(ot_instance, OT_PANID);
      |                 ^~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:493:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  493 |                 otThreadSetExtendedPanId(ot_instance, &xpanid);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:498:4: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  498 |                         otThreadSetNetworkKey(ot_instance, &networkKey);
      |                         ^~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/subsys/net/l2/openthread/openthread.c:568:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  568 |                 otPlatUartEnable();
      |                 ^~~~~~~~~~~~~~~~
10 warnings generated.
```

**Expected behavior**
No warnings or errors.

**Impact**
Uncertain. OpenThread seems to have reasons to want the caller to check return value and when caller doesn't do this then it's hard to predict what may fail later.

**Logs and console output**
See warnings above.

**Environment (please complete the following information):**

 - OS: Linux
 - Toolchain : LLVM (LLVM for Arm)
 - Commit SHA or Version used: a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3

**Additional context**
An easy fix would be to use `(void)` or disable the warning for clang, but based on the effort carried out by OpenThread maintainers to actually request the return value then this seems like a bad idea.

This is also the reason why this is opened as a bug and not an enhancement.

Link to locations in code base:
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L341
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L442
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L453
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L454
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L489
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L490
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L491
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L493
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L498
https://github.com/zephyrproject-rtos/zephyr/blob/a1a92a20a60ecb66bc37c931133d18bc5fd2c1f3/subsys/net/l2/openthread/openthread.c#L568

--- a/subsys/net/l2/openthread/openthread.c
+++ b/subsys/net/l2/openthread/openthread.c
@@ -338,7 +338,11 @@ static void ot_joiner_start_handler(otError error, void *context)
 	switch (error) {
 	case OT_ERROR_NONE:
 		NET_INFO("Join success");
-		otThreadSetEnabled(ot_context->instance, true);
+		error = otThreadSetEnabled(ot_context->instance, true);
+		if (error != OT_ERROR_NONE) {
+			NET_ERR("Failed to start the OpenThread network [%d]", error);
+		}
+
 		break;
 	default:
 		NET_ERR("Join failed [%d]", error);
@@ -439,7 +443,11 @@ int openthread_start(struct openthread_context *ot_context)
 		goto exit;
 	}
 
-	otIp6SetEnabled(ot_context->instance, true);
+	error = otIp6SetEnabled(ot_context->instance, true);
+	if (error != OT_ERROR_NONE) {
+		NET_ERR("Failed to set %s [%d]", "IPv6 support", error);
+		goto exit;
+	}
 
 	/* Sleepy End Device specific configuration. */
 	if (IS_ENABLED(CONFIG_OPENTHREAD_MTD_SED)) {
@@ -450,8 +458,17 @@ int openthread_start(struct openthread_context *ot_context)
 		 */
 		ot_mode.mRxOnWhenIdle = false;
 
-		otThreadSetLinkMode(ot_context->instance, ot_mode);
-		otLinkSetPollPeriod(ot_context->instance, OT_POLL_PERIOD);
+		error = otThreadSetLinkMode(ot_context->instance, ot_mode);
+		if (error != OT_ERROR_NONE) {
+			NET_ERR("Failed to set %s [%d]", "link mode", error);
+			goto exit;
+		}
+
+		error = otLinkSetPollPeriod(ot_context->instance, OT_POLL_PERIOD);
+		if (error != OT_ERROR_NONE) {
+			NET_ERR("Failed to set %s [%d]", "poll period", error);
+			goto exit;
+		}
 	}
 
 	/* Configure Child Supervision and MLE Child timeouts. */
@@ -486,16 +503,39 @@ int openthread_start(struct openthread_context *ot_context)
 		otExtendedPanId xpanid;
 		otNetworkKey    networkKey;
 
-		otThreadSetNetworkName(ot_instance, OT_NETWORK_NAME);
-		otLinkSetChannel(ot_instance, OT_CHANNEL);
-		otLinkSetPanId(ot_instance, OT_PANID);
+		error = otThreadSetNetworkName(ot_instance, OT_NETWORK_NAME);
+		if (error != OT_ERROR_NONE) {
+			NET_ERR("Failed to set %s [%d]", "network name", error);
+			goto exit;
+		}
+
+		error = otLinkSetChannel(ot_instance, OT_CHANNEL);
+		if (error != OT_ERROR_NONE) {
+			NET_ERR("Failed to set %s [%d]", "channel", error);
+			goto exit;
+		}
+
+		error = otLinkSetPanId(ot_instance, OT_PANID);
+		if (error != OT_ERROR_NONE) {
+			NET_ERR("Failed to set %s [%d]", "PAN ID", error);
+			goto exit;
+		}
+
 		net_bytes_from_str(xpanid.m8, 8, (char *)OT_XPANID);
-		otThreadSetExtendedPanId(ot_instance, &xpanid);
+		error = otThreadSetExtendedPanId(ot_instance, &xpanid);
+		if (error != OT_ERROR_NONE) {
+			NET_ERR("Failed to set %s [%d]", "ext PAN ID", error);
+			goto exit;
+		}
 
 		if (strlen(OT_NETWORKKEY)) {
 			net_bytes_from_str(networkKey.m8, OT_NETWORK_KEY_SIZE,
 					   (char *)OT_NETWORKKEY);
-			otThreadSetNetworkKey(ot_instance, &networkKey);
+			error = otThreadSetNetworkKey(ot_instance, &networkKey);
+			if (error != OT_ERROR_NONE) {
+				NET_ERR("Failed to set %s [%d]", "network key", error);
+				goto exit;
+			}
 		}
 	}
 
@@ -542,7 +582,7 @@ static int openthread_init(struct net_if *iface)
 		.name = "openthread",
 		.no_yield = true,
 	};
-	otError err;
+	otError err = OT_ERROR_NONE;
 
 	NET_DBG("openthread_init");
 
@@ -565,7 +605,11 @@ static int openthread_init(struct net_if *iface)
 	}
 
 	if (IS_ENABLED(CONFIG_OPENTHREAD_COPROCESSOR)) {
-		otPlatUartEnable();
+		err = otPlatUartEnable();
+		if (err != OT_ERROR_NONE) {
+			NET_ERR("Failed to enable UART: [%d]", err);
+		}
+
 		otNcpHdlcInit(ot_context->instance, ncp_hdlc_send);
 	} else {
 		otIp6SetReceiveFilterEnabled(ot_context->instance, true);
@@ -613,7 +657,7 @@ static int openthread_init(struct net_if *iface)
 
 	(void)k_work_submit_to_queue(&ot_context->work_q, &ot_context->api_work);
 
-	return 0;
+	return (err == OT_ERROR_NONE) ? 0 : -EIO;
 }
 
 void ieee802154_init(struct net_if *iface)

